"""
The main file of this project. All the rules will be imported from seperate files.
Goal: 
    Create a de novo assembly from paired end sequences with quality assesments
Author: 
    Mark van de Streek
Date: 
    2024-03-16
Copyright: 
    BFV3 2023-2024

Run pipeline:
snakemake -c 1 --use-conda --conda-frontend conda
"""

configfile: "config/config.yaml",
workdir: config['workdir'],

rule all:
    input:
        # The FASTQC Tool output isn't used in any rule,
        # therefore, we must specify this in the rule all.
        expand("results/fastqc/{sample}_1_fastqc.html", sample=config['samples']),
        expand("results/fastqc/{sample}_2_fastqc.html", sample=config['samples']),
        expand("results/fastx/filtered_{sample}_1.fastq", sample=config['samples']),
        expand("results/fastx/filtered_{sample}_2.fastq", sample=config['samples']),
        # The "last" file is the output of the pipeline
        nonRNA=expand("results/sortmerna/{sample}_non_rRNA.fq", sample=config['samples']),
        alignedRNA=expand("results/sortmerna/{sample}_rRNA.fq", sample=config['samples']),

rule fastQC:
    input:
        config["samples_directory"] + "{sample}_1.fastq",
        config["samples_directory"] + "{sample}_2.fastq"
    output:
        "results/fastqc/{sample}_1_fastqc.html",
        "results/fastqc/{sample}_1_fastqc.zip",
        "results/fastqc/{sample}_2_fastqc.html",
        "results/fastqc/{sample}_2_fastqc.zip",
    message:
        "Running FastQC on {input}"
    conda:
        "../envs/FASTQC.yaml"
    params:
        threads=config['threads']
    shell:
        "fastqc -t {params.threads} {input} -o results/fastqc/"

rule fastx_toolkit:
    """
    This tool has verbose specified, but runs first all operations, 
    and then prints the output.
    """
    input:
        s1=config["samples_directory"] + "{sample}_1.fastq",
        s2=config["samples_directory"] + "{sample}_2.fastq"
    output:
        s1="results/fastx/filtered_{sample}_1.fastq",
        s2="results/fastx/filtered_{sample}_2.fastq"
    message:
        "Running fastx_toolkit on {input}"
    conda:
        "../envs/FASTX_TOOLKIT.yaml"
    shell:
        """
        fastq_quality_filter -v -q 20 -p 75 -i {input.s1} -o {output.s1}
        fastq_quality_filter -v -q 20 -p 75 -i {input.s2} -o {output.s2}
        """

rule trimmomatic_pe:
    input:
        r1="results/fastx/filtered_{sample}_1.fastq",
        r2="results/fastx/filtered_{sample}_2.fastq"
    output:
        r1="results/trimmomatic/{sample}_1_paired_trimmed.fastq",
        r2="results/trimmomatic/{sample}_2_paired_trimmed.fastq",
        r1_unpaired="results/trimmomatic/{sample}_1_unpaired_trimmed.fastq",
        r2_unpaired="results/trimmomatic/{sample}_2_unpaired_trimmed.fastq"
    log:
        "logs/trimmomatic/{sample}.log"
    params:
        # Specify the trimmer options
        trimmer=[
            "ILLUMINACLIP:../envs/trimmomatic-0.39/adapters/TruSeq3-PE.fa:2:30:10",
            "SLIDINGWINDOW:5:20",
            "LEADING:5",
            "TRAILING:5",
            "MINLEN:50"],

        extra="",
        compression_level="-9"
    threads: 16
    resources:
        mem_mb=1024
    wrapper:
        "v3.5.2/bio/trimmomatic/pe"

rule mergeFastQ:
    input:
        r1="results/trimmomatic/{sample}_1_paired_trimmed.fastq",
        r2="results/trimmomatic/{sample}_2_paired_trimmed.fastq"
    output:
        "results/sortmerna/{sample}_interleaved.fastq"
    shell:
        "workflow/scripts/merge-paired-reads.sh {input.r1} {input.r2} {output}"

rule sortMeRNA:
    """
    SortMeRNA uses a Work directory with the default location $USER/sortmerna/run/. Sortmerna also checks kvdb directory. 
    This directory has to be Empty prior each new run. Therefore, we have to empty this before starting a new run.
    """
    input:
        ref1=config['rRNA_databases'] + "silva-bac-16s-id90.fasta", 
        ref2=config['rRNA_databases'] + "silva-bac-23s-id98.fasta",
        ref3=config['rRNA_databases'] + "silva-arc-16s-id95.fasta", 
        ref4=config['rRNA_databases'] + "silva-arc-23s-id98.fasta",
        ref5=config['rRNA_databases'] + "silva-euk-18s-id95.fasta",
        ref6=config['rRNA_databases'] + "silva-euk-28s-id98.fasta",
        ref7=config['rRNA_databases'] + "rfam-5s-database-id98.fasta", 
        ref8=config['rRNA_databases'] + "rfam-5.8s-database-id98.fasta",
        sample="results/sortmerna/{sample}_interleaved.fastq",
    output:
        nonRNA="results/sortmerna/{sample}_non_rRNA.fq",
        alignedRNA="results/sortmerna/{sample}_rRNA.fq",
        #
        # PROBLEM: Tool has standard as output .fq, but when I say output is .fq, 
        # Output is .fq.fq. So I need to change the output to .fq, but then the
        # output is not recognized as a file.
        #
    conda:
        "../envs/sortmerna.yaml",
    threads: 
        threads=config['threads'],
    log:
        "logs/sortmerna/{sample}.log",
    shell:
        """
        rm -rf ~/sortmerna/run/kvdb/*
        sortmerna --reads {input.sample} --ref {input.ref1} --ref {input.ref2} --ref {input.ref3} --num_alignments 1 --fastx --aligned {output.alignedRNA} --other {output.nonRNA} -a {threads} -m 64000 --paired_in -v
        mv {output.alignedRNA}.fq {output.alignedRNA} &&
        mv {output.nonRNA}.fq {output.nonRNA}
        """
